### ISSUE TITLE
[Feature Request] add checksum tag for releases api

### ISSUE THREAD

**Issue by LazyT** [go-gitea/gitea#15694]
Hi,

I use the [API](https://try.gitea.io/api/swagger#/repository/repoListReleases) to check for new versions and download them directly in my program.

At the moment it's unfortunately only possible to check the file size via "size:xxx". An additional crc tag like "md5:xxx" for each file would be safer to validate the download.

Could this be implemented please?

Thank you!

ref: https://codeberg.org/Codeberg/Community/issues/435

**Comment by 6543** [comment]
@LazyT thanks for opening the issue.

the question is if md5 should be used because of collisions issues ...
[Signature/Hash Choice for File Integrity Verification](https://stackoverflow.com/questions/13269606/signature-hash-choice-for-file-integrity-verification)

**Comment by LazyT** [comment]
Add several tags "md5:xxx", "sha256:xxx", "sha512:xxx" so anybody can choose?

For me all hashes supported by [Qt](https://doc.qt.io/qt-5/qcryptographichash.html#Algorithm-enum) are fine. :smirk:

**Comment by LazyT** [comment]
> Or maybe the thing you are trying to accomplish is to only check against download-errors, where I'm not sure if they could even occur

Why do download sites provide an checksum for the files? Would you flash an bios without verifying the file after download? I'm not. :wink:

For evil attackers the files should better use a signature.

**Comment by naums** [comment]
TCP does that for you. The only thing why a file might get corrupted could be, that your PC crashed before the file is written from the OS out of the page cache to disk.

Nonetheless, unencrypted checksums together with the file is virtually useless against an attacker switching the file. Linux Distros use checksums always together with crypto to check against switched files (and in turn connection problems). I wouldn't flash a BIOS chip at all, especially when I can't be sure who put the file there.

When doing checksumming, placing the burden to your user, i.e. the programmer providing the binary, is a bit weird. A programmer wants to compile her code as always and the system(s) (CI, Gitea) should add crypto transparent to the programmer. Don't trust the user to do it right, that never works out.

tl;dr: Think about what you want to achieve, if it is a real problem and whether your solution suffices to solve the problem.

**Comment by LazyT** [comment]
> Think about what you want to achieve

I think it's clear what I want to achieve: simply verify the update (automatically in my app, not by user) after the download is finished.

**Comment by naums** [comment]
Verify against what? That the file is the same on your PC as is on the server - yes, your checksum works. Verify against someone with FTP-access replacing the file with malware - will not be secure without a sound idea about encryption.

My point is: a checksum is not really useful, you want a signature.

**Comment by LazyT** [comment]
> That the file is the same on your PC as is on the server - yes, your checksum works.

Great, that's what I want.

> you want a signature.

No, as written above I would use a signature in my app to protect against evil hackers. That has nothing to do with the checksum to verify the download is corrupt...
